### DISCUSSION TITLE
Geometry：geometryStream stored in .bim sqlite database （StandaloneDb）

### DISCUSSION THREAD

**Discussion by QianSunCN** [iTwin/itwinjs-core#5766] [Q&A] [answered]
hi, when I insert an indexedPolyface to StandaloneDb, then use the ecsql to query geometryStream from BisCore.GeometricElement3d, for example, “select geometryStream geom from BisCore.GeometricElement3d WHERE ECInstanceId=0x17”.

And, I get the geometryStream with Binary format, I'm curious about how these data were generated.

I also use the method, for example, "const by64 = BentleyGeometryFlatBuffer.geometryToBytes(indexedPolyface, true)"
I found that “by64 ” and “geom” are different, what is the reason for this problem?

thank you!

----------------
the "geom" is like:
Uint8Array(132) [
  132,   0, 232,   1, 48,  0,  6,  0,   0, 224,  0,  0,
    0,   1,   0,   0,  0,  8, 13,  8,   1,   1, 64,  9,
    0,   0,   0, 200,  0,  0,  0, 98, 103,  48, 48, 48,
   49, 102,  98,  20,  1, 21,  5,  1,  16,  10,  0, 14,
    0,   7,   5,  44,  0, 10,  5, 16,   8,   0, 13, 48,
    5,   8,   8,  42,  0, 20,  5, 22,   9,   1,  0,  4,
    9,   7,  13,   1,  8, 16,  0, 15,  13,  10,  1,  1,
   32,  42,   0,   0,  0, 16,  0,  0,   0,  32,  1, 13,
   20,   0,   0,   1,
  ... 32 more items
]

the "by64 " is like:
Uint8Array(176) [
  98, 103, 48, 48, 48, 49, 102, 98, 16,  0,  0, 0,
   0,   0,  0,  0,  8,  0,  10,  0,  9,  0,  4, 0,
   8,   0,  0,  0, 36,  0,   0,  0,  0, 13, 30, 0,
  16,   0, 12,  0,  0,  0,   0,  0,  0,  0,  0, 0,
   8,   0,  0,  0,  0,  0,   0,  0,  0,  0,  0, 0,
   0,   0,  4,  0, 30,  0,   0,  0,  1,  0,  0, 0,
   8,   0,  0,  0, 24,  0,   0,  0,  4,  0,  0, 0,
   1,   0,  0,  0,  2,  0,   0,  0,  3,  0,  0, 0,
   0,   0,  0,  0,
  ... 76 more items
]

**Comment by bbastings** [comment]
The geometry stream is stored compressed.

Use [ElementGeometry.Iterator](https://www.itwinjs.org/reference/core-common/geometry/elementgeometry/) in conjunction with [IModelDb.elementGeometryRequest](https://www.itwinjs.org/reference/core-backend/imodels/imodeldb/elementgeometryrequest/) to extract and iterate over a geometric element's geometry stream.

**Reply by QianSunCN** [reply]
Thank you very much for your answer, it works, I successfully extracted element's geometry stream by your method, the result just like:

IndexedPolyface {
  geometryCategory: 'polyface',
  data: PolyfaceData {
    face: [],
    point: GrowableXYZArray {
      _data: [Float64Array],
      _xyzInUse: 3,
      _xyzCapacity: 8,
      _growthFactor: 1.5
    },
    pointIndex: [ 0, 1, 2 ],
    edgeVisible: [ true, true, true ],
    _twoSided: true,
    _expectedClosure: 0,
    auxData: undefined
  },
  _facetStart: [ 0, 3 ],
  _facetToFaceData: []
}

You say the geometry stream is stored compressed, could you show me what algorithm is used to implement this compression process.
In short, how to transform the above-mentioned "by64 " to "geom"?

**Reply by pmconne** [reply, marked as answer]
It's compressed using [Snappy](https://en.wikipedia.org/wiki/Snappy_(compression)).
[Relevant code](https://github.com/iTwin/imodel-native/blob/29365d4683eb5d08fa7d6eaeaad7a5d071f62a46/iModelCore/iModelPlatform/DgnCore/DgnElement.cpp#L3703).

**Reply by QianSunCN** [reply]
thanks，please could tell me the repository（https://github.com/iTwin/imodel-native ） is the specific implementation code for database (.bim) reading and writing or others?

**Reply by pmconne** [reply]
That repository is used to produce the [Node addon](https://www.npmjs.com/package/@bentley/imodeljs-native) consumed by the [@itwin/core-backend](https://www.npmjs.com/package/@itwin/core-backend) package.
@itwin/core-backend provides the [APIs](https://www.itwinjs.org/learning/backend/) for reading and writing iModels.

**Reply by QianSunCN** [reply]
thanks
